### PR TITLE
Decouple scalar field calculation from color rendering

### PR DESCRIPTION
**Pull request by MatthewPeterKelly**
## Summary:

Major refactor for the rendering pipeline to decouple the "field evaluation" and "color mapping". This is done so that (in a future commit) we can dynamically change the color map without needing to recompute the underlying scalar field. This makes it much faster to interactively iterate on the color of the fractal, even at full resolution.

This PR is "phase two" of the larger roadmap plan outlined in `docs/gui-unification-roadmap.md`. The code for the three primary commits (details below) was written by Claude Opus 4.7, based on extensive interactive planning.

Follow-up to https://github.com/MatthewPeterKelly/fractal-renderer/pull/167.

Co-Authored-By: Claude Opus 4.7 (1M context) <noreply@anthropic.com>

## Detailed Changes:

**commit 4199c23ab5730441afc6cfa9d02c7f98ecc90afc**

    feat: add ColorMapKind + RenderingPipeline (Phase 2.1, parallel-to-old)
    
    Phase 2 commit 2.1: introduce the new compute / color split machinery
    end-to-end without disturbing the production runtime.
    
    - Add `ColorMapKind` trait with statically-typed `Cell` and `Cache`
      associated types. Per-variant impls on `ForegroundBackground`,
      `BackgroundWithColorMap`, and `MultiColorMap`. Cache is rebuilt in
      place from keyframes via `refresh_cache` (allocation-free), and
      `colorize_cell` is a no-`dyn` per-(sub)pixel hot-path lookup.
    - Extend `Renderable` with `type ColorMap`, `compute_raw_field`,
      `populate_histogram`, `normalize_field`, and `color_map`. The
      histogram and normalize methods default to no-ops (DDP path).
    - Add `src/core/render_pipeline.rs` with `RenderingPipeline<F>`
      orchestrating the four phases (compute_raw_field → populate_histogram
      → normalize_field → refresh_cache + colorize_collapse) against
      preallocated buffers. The generic free function `colorize_collapse`
      walks the row-major output `egui::ColorImage`, averaging `(n+1)²`
      subpixel `[u8;3]` results per output pixel; fully monomorphized over
      the `ColorMapKind`.
    - Implement the new methods on `QuadraticMap<T>`, `DrivenDampedPendulumParams`,
      and `NewtonsMethodRenderable<F>`. Histogram source for the gradient
      fractals continues to use the legacy sub-sample grid in 2.1; 2.3
      switches to a full-field walk.
    - Rename `QuadraticMapParams::color_map` -> `color_map_params` so the
      renderer-level `color_map()` does not shadow the params accessor;
      touches mandelbrot.rs, julia.rs, benches/benchmark.rs.
    - Add unit tests on each `ColorMapKind` impl (None / endpoint / mid /
      out-of-range root index / refresh-on-edit) and on `colorize_collapse`
      (synthetic 4x4 AA averaging, no-AA single-cell, gradient endpoints).
    - Add `tests/phase_2_pixel_equivalence_tests.rs` exercising the new
      pipeline against the legacy `Renderable::render_to_buffer` path on
      small Mandelbrot/Julia/Newton/DDP fixtures. DDP asserts strict
      equality (no lookup table); the gradient fractals assert per-channel
      diffs within `MAX_GRADIENT_DIFF=32`. The looser tolerance reflects an
      intentional architectural change: the legacy LUT is built over
      `[cdf.min_data, cdf.max_data]` with the CDF baked into each entry,
      while the new cache is built over `[0, 1]` with the CDF applied per
      cell in `normalize_field`. The latter shape is a hard requirement for
      Phase 6 re-colorize-only (cache stays valid when the CDF doesn't
      change), but the two quantization schemes can disagree by a small
      amount near steep gradient transitions. Phase 2.2 will regenerate
      pixel-hash regression fixtures accordingly.
    - New machinery is `#[allow(dead_code)]` until 2.2 wires it into the
      production runtime.
    
   
**commit 0caa21cbf167701ca136f945f3eaa8ba120bdb4b**

    feat: switch runtime to RenderingPipeline; collapse render_options to sampling_level
    
    Phase 2 commit 2.2: production runtime now goes through
    `RenderingPipeline`. The legacy compute/colorize path is removed.
    
    - `RenderOptions` collapses `subpixel_antialiasing: u32` and
      `downsample_stride: usize` into a single signed `sampling_level: i32`.
      Positive = AA, 0 = baseline, negative = block-fill (nearest-neighbor;
      the legacy `KeyframeLinearPixelInerpolation` interpolation is dropped).
      `SpeedOptimizer for RenderOptions` mutates the unified axis.
    - `image_utils::render` (offline render) and `PixelGrid` (interactive)
      both go through `RenderingPipeline::render`. `Renderable::render_point`
      and the default `render_to_buffer` are deleted; each fractal's impl
      now exposes `histogram_bin_count`, `histogram_max_value`,
      `lookup_table_count`, plus the `compute_raw_field` /
      `populate_histogram` / `normalize_field` / `color_map` methods added in
      2.1. `compute_raw_field` (and the colorize/normalize walks) handles
      positive AA, baseline, and negative block-fill against a field sized
      for the user's max sampling level.
    - `QuadraticMap<T>` slims down to just `{ fractal_params }`. Histogram,
      CDF, and color cache live in the pipeline.
      `NewtonsMethodRenderable<F>` slims down to `{ params, system }`.
    - Delete `generate_scalar_image*`, `wrap_renderer_with_antialiasing`,
      `KeyframeLinearPixelInerpolation`, `render_image_internal`,
      `fill_skipped_entries`, `render_single_row_within_image`,
      `PixelRenderLambda`, `subpixel_offset_vector`, `create_buffer`,
      `create_empty_histogram`, and `src/fractals/utilities.rs`. The
      `display_buffer_to_color_image` transposition is gone too — the
      pipeline writes directly into the row-major `egui::ColorImage`.
      `SubpixelGridMask` / `UpsampledPixelMapper` / `SubpixelIndex` are
      retained for `chaos_game` (out of scope per roadmap §9.5).
    - Benchmark migrated to drive `RenderingPipeline::render` end-to-end.
    - All 39 example / regression / bench JSON files migrated to
      `sampling_level` via `scripts/migrate_phase_2_render_options.py`
      (committed alongside).
    - Pixel-hash regression fixtures regenerated for Mandelbrot, Julia,
      and the downsample fixture. Two distinct sources of pixel drift:
      (1) The new color cache is indexed over `[0, 1]` with the CDF applied
          per cell in `normalize_field`, while the legacy lookup was over
          `[cdf.min_data, cdf.max_data]` with the CDF baked into each table
          entry. The two quantization schemes round to different bins near
          steep gradient transitions.
      (2) Block-fill is nearest-neighbor; the legacy path used bilinear
          interpolation between sparse samples.
      DDP, Barnsley, and Sierpinski hashes are unchanged. The
      `phase_2_pixel_equivalence_tests` suite (introduced in 2.1) is
      removed alongside the legacy path it was comparing against.
    
**commit 8008aff46d3deac0147b6166a96c4cf7e1e12fd8**

    feat: histogram source from full field; drop histogram_sample_count
    
    Phase 2 commit 2.3. The histogram that drives gradient normalization
    for Mandelbrot, Julia, and Newton now reflects the same population the
    colorize pass reads, instead of an independent sub-sample grid.
    
    - `Renderable::populate_histogram` for `QuadraticMap` and
      `NewtonsMethodRenderable` switched to full-field walks via shared
      `walk_populated_*_cells` helpers (read-only siblings of the
      compute/normalize walks introduced in 2.2). Newton's histogram now
      uses smooth iteration counts instead of the integer iteration count.
    - `histogram_sample_count` is removed from `ColorMapParams` and
      `CommonParams`, and from the QuadraticMap/Newton speed-optimizer
      reference caches. The field is now silently ignored by serde so the
      existing JSONs (which still carry it) continue to deserialize. A
      follow-up cleanup commit can strip the dead field.
    - Drop `ImageSpecification::scale_to_total_pixel_count` (sub-sample
      grid was its only caller). Tag `NewtonRhapsonResult.iteration_count`
      and the `QuadraticMapParams::color_map_params_mut` trait method as
      `#[allow(dead_code)]` since they're public surface kept for Phase 6
      / future diagnostics. DDP / Barnsley / Sierpinski paths unaffected.
    - Pixel-hash regression fixtures regenerated for Mandelbrot, Julia,
      and the downsample fixture. The hash diff vs. 2.2 reflects the
      histogram source change (not a structural rendering change); spot
      checking the rendered PNGs confirms the output looks visually
      consistent with the 2.2 outputs at the same parameters. DDP /
      Barnsley / Sierpinski hashes unchanged.

### CODE REVIEW

File: docs/phase-2-detailed-plan.md
Comment (MatthewPeterKelly) [inline docs/phase-2-detailed-plan.md:1]: Remove this file before committing.

File: docs/gui-unification-roadmap.md
Comment (MatthewPeterKelly) [inline docs/gui-unification-roadmap.md:1]: Updates LGTM.

File: examples/common/mod.rs
Comment (MatthewPeterKelly) [inline examples/common/mod.rs:1]: Manually verified (smoke test) that all interactive examples run (render, explore, color).

File: src/core/color_map.rs [hunk 1]
```diff
@@ -170,6 +173,159 @@ impl ColorMapper for ColorMapLookUpTable {
...
+    /// In-place rebuild of the cache from current keyframes / flat colors.
+    /// Allocation-free.
+    fn refresh_cache(&self, cache: &mut Self::Cache);
+
+    /// Per-cell color lookup. Statically dispatched; called inside the
+    /// AA-collapse loop. No allocation, no `dyn`.
+    fn colorize_cell(cache: &Self::Cache, cell: Self::Cell) -> [u8; 3];
+}
+
+fn color32_to_rgb(c: Color32) -> [u8; 3] {
```
Comment (MatthewPeterKelly) [inline src/core/color_map.rs:208]: Neat.  EGUI for the win.  Maybe eventually we should use `Color32` across the library. [u8, 3] is fine for now.

File: src/fractals/driven_damped_pendulum.rs [hunk 2]
```diff
@@ -68,20 +57,96 @@ impl Renderable for DrivenDampedPendulumParams {
...
+                    *cell = compute_basin_of_attraction(
+                        &[re, im],
+                        time_phase,
+                        n_max_period,
+                        n_steps_per_period,
+                        tol,
+                    );
+                }
+            });
+        }
```
Comment (MatthewPeterKelly) [inline src/fractals/driven_damped_pendulum.rs:149]: This is no good!  The `sampling_level` logic needs to be at the top level, common to all fractals. Moving it down into a specific fractal like this forks implementation and makes it very difficult to add and maintain new fractals.

File: src/fractals/newtons_method.rs [hunk 3]
```diff
@@ -353,31 +303,189 @@ where
...
+                        continue;
+                    }
+                    let block_y = outer_y / stride;
+                    let py = (block_y * block_size) as u32;
+                    let im = pixel_map.height.map(py);
+                    *cell = evaluate(re, im);
+                }
+            });
+        }
+    }
```
Comment (MatthewPeterKelly) [inline src/fractals/newtons_method.rs:389]: Same problem here - duplicated logic and it is implemented per-fractal. This needs to be hoisted up into the top-level libraries. The per-fractal code should be simple, and just concerned with evaluating the "scalar field" at a single point.

File: src/fractals/quadratic_map.rs [hunk 4]
```diff
@@ -287,39 +245,235 @@ where
...
+                    continue;
+                }
+                let block_y = outer_y / stride;
+                let py = (block_y * block_size) as u32;
+                let im = pixel_map.height.map(py);
+                *cell = escape_count(&[re, im]);
+            }
+        });
+    }
+}
```
Comment (MatthewPeterKelly) [inline src/fractals/quadratic_map.rs:388]: Same problem.

**Review by MatthewPeterKelly — COMMENTED** [review]
Close!  Traits mostly look good, but the antialiasing logic is implemented on a per-fractal basis, rather than in the core library. Rework time.

**Comment by MatthewPeterKelly** [comment]
Closing in favor of :  https://github.com/MatthewPeterKelly/fractal-renderer/pull/170
